ledger-api/grpc-definitions/com/daml/ledger/api/v1/event_query_service.proto

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+syntax = "proto3";
+
+package com.daml.ledger.api.v1;
+
+import "com/daml/ledger/api/v1/event.proto";
+import "com/daml/ledger/api/v1/value.proto";
+
+option java_outer_classname = "EventQueryServiceOuterClass";
+option java_package = "com.daml.ledger.api.v1";
+option csharp_namespace = "Com.Daml.Ledger.Api.V1";
+
+// Allow clients to query events
+service EventQueryService {
+
+  // Get the create and the consuming exercise event for the contract with the provided ID
+  rpc GetEventsByContractId (GetEventsByContractIdRequest) returns (GetEventsByContractIdResponse);
+
+  // Get all create and consuming exercise events for the contracts with the provided contract key
+  rpc GetEventsByContractKey (GetEventsByContractKeyRequest) returns (GetEventsByContractKeyResponse);
+
+}
+
+message GetEventsByContractIdRequest {
+  // The contract id being queried.
+  // Required
+  string contract_id = 1;
+
+  // The parties whose events the client expects to see.
+  // The events associated with the contract id will only be returned if the requesting parties includes
+  // at least one party that is a witness of the action that created the event. For a full description
+  // of witness see https://docs.daml.com/concepts/ledger-model/ledger-privacy.html#projections.
+  // Required
+  repeated string requesting_parties = 2;
+
+}
+
+message GetEventsByContractIdResponse {
+
+  // Will be populated if a created event is found for the contract id
+  // Optional
+  CreatedEvent create_event = 1;
+
+  // Will be populated if a archive event is found for the contract id
+  // Optional
+  ArchivedEvent archive_event = 2;
+
+}
+
+message GetEventsByContractKeyRequest {
+
+  // The contract key to search for.
+  // Required
+  Value contract_key = 1;
+
+  // The template id associated with the contract key
+  // Required
+  Identifier template_id = 2;
+
+  // The parties whose events the client expects to see.
+  // The events associated with the contract key will only be returned if the requesting parties includes
```

Comment:
If you include witnesses rather than filter by stakeholders you can get in a situation where you see two creates without an intermediate archive in between. What will I get as a result?

Reply:
Agree, this is consusing, switching to stakeholders.